hashing/src/indexed_merkle_proof.rs

```diff
@@ -326,17 +326,17 @@ mod test {
     #[test]
     fn empty_proof() {
         let empty_merkle_root = Digest::hash_merkle_tree(vec![]);
-        assert_eq!(
-            empty_merkle_root,
-            Digest::hash_pair(0u64.to_le_bytes(), Digest::SENTINEL_MERKLE_TREE)
-        );
+        assert_eq!(empty_merkle_root, Digest::SENTINEL_MERKLE_TREE);
         let indexed_merkle_proof = IndexedMerkleProof {
             index: 0,
             count: 0,
             merkle_proof: vec![],
         };
         assert_eq!(indexed_merkle_proof.verify(), Ok(()));
```

Comment:
Why is this `Ok`? If `index >= count` it should fail, shouldn't it?

Reply:
Hmm, good catch.
What I think happened is:
1. Once upon a time [we had an implementation](https://github.com/xcthulhu/casper-node/blob/chunked-tries/hashing/src/primitives.rs#L165-L219) of the `IndexedMerkleProof::new()` that allowed empty data and index 0. As a result of `new(vec![], 0)`, it returned the following:
```
Ok(
    IndexedMerkleProof {
        index: 0,
        count: 0,
        merkle_proof: [],
    },
)
```

2. However, for the same call current implementation returns:
```
Err(
    IndexOutOfBounds {
        count: 0,
        index: 0,
    },
)
```

To work correctly with the [previous implementation](https://github.com/xcthulhu/casper-node/blob/chunked-tries/hashing/src/primitives.rs#L313), the `verify()` function initially had the following check, which was not removed when I changed the behavior of the constructor:
```
if !((self.count == 0 && self.index == 0) || self.index < self.count) {
```
which now should look like this:
```
if self.index >= self.count {
```

That will imply changes in the following tests:
```
empty_proof
empty_unexpected_proof_length
```

@afck @xcthulhu your opinion?

Reply:
I agree, `verify` needs to be updated: `index >= count` is _never_ valid.

Reply:
Updated in https://github.com/casper-network/casper-node/pull/2119/commits/a1854dee87fe3367135579f7cb9be661a50540e4